storage/internal/benchmarks/main.go

```diff
@@ -249,28 +251,53 @@ func main() {
 	benchGroup, _ := errgroup.WithContext(ctx)
 	benchGroup.SetLimit(opts.numWorkers)
 
+	exitWithErrorCode := false
+
 	// Run benchmarks
 	for i := 0; i < opts.numSamples && time.Since(start) < opts.timeout; i++ {
 		benchGroup.Go(func() error {
 			benchmark := w1r3{opts: opts, bucketName: opts.bucket}
-			if err := benchmark.setup(); err != nil {
-				log.Fatalf("run setup failed: %v", err)
+			if err := benchmark.setup(ctx); err != nil {
+				// If setup failed once, it will probably continue failing.
+				// Returning the error here will cancel the context to stop the
+				// benchmarking.
+				return fmt.Errorf("run setup failed: %v", err)
 			}
 			if err := benchmark.run(ctx); err != nil {
-				log.Fatalf("run failed: %v", err)
+				// If a run fails, we continue, as it could be a temporary issue.
+				// We log the error and make sure the program exits with an error
+				// to indicate that we did see an error, even though we continue.
+				log.Printf("run failed: %v", err)
+				exitWithErrorCode = true
+			}
+			if err := benchmark.cleanup(); err != nil {
+				// If cleanup fails, we continue, as a single fail is not critical.
+				// We log the error and make sure the program exits with an error
+				// to indicate that we did see an error, even though we continue.
+				// Cleanup may be expected to fail if there is an issue with the run.
+				log.Printf("run cleanup failed: %v", err)
+				exitWithErrorCode = true
 			}
 			return nil
 		})
 	}
 
-	benchGroup.Wait()
+	err := benchGroup.Wait()
 	close(results)
 	recordResultGroup.Wait()
 
 	if outputFile != "" {
 		// if sending output to a file, we can use stdout for informational logs
 		fmt.Printf("\nTotal time running: %s\n", time.Since(start).Round(time.Second))
 	}
+
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	if exitWithErrorCode {
```

Comment:
Under what circumstances would `exitWithErrorCode` be true but the script would not have already exited on L295 above?

Reply:
So for runs and cleanup, if you see L270 and L278, we log the error and set `exitWithErrorCode`, but we do not return the error. This is because we want to continue benchmarking, but still log the error. If you return the error, the group's context is cancelled. 

I just added a commit (missed that earlier) to make sure the group's context (based on the original context) is the one being used in the setup and runs. So, any future runs would stop on seeing the context cancelled when an error is returned. We want this for setup since a setup error indicates in most cases that future setups will also fail.

So, if setup returns a failure, runs would be cancelled, cleanup done, and then `Errgroup.Wait` would return with an error and the script would exit on L295.

If the run or cleanup returns a failure, it would be logged, and the script would continue until it finishes `i` samples, then `Errgroup.Wait` would return without an error and we would reach these lines indicating to exit with an error code.